Basic scan and mirror support

SirTyson opened the pull request:
This is the first of several PRs implementing stellar-archivist functionality in Rust. In particular, this PR adds basic support for the mirror and scan operations, the control flow of stellar-archivist, and unit tests.

Feature wise, this supports mirroring and scanning archives without verification from filesystem and HTTP/HTTPS backend. It supports the options `--low`, `--high`, `--skip-optional`, and `-c` similar to the [go tool](https://pkg.go.dev/github.com/stellar/go/tools/stellar-archivist#section-readme). Most notably, this does not yet support repair, `--thorough`, or `--verify`, which are coming in follow up PRs.

Mostly, the tool is feature equivalent to the go tool, with a few differences (other than the obviously in flight features):

- Maintains .well-known/stellar-history.json files, meaning a call to `mirror` will initialize the .well-known file if it does not exist, or update it if it does. `scan` also does validity checking on this file.
- Safeguards around gap detection. If a certain mirror command will result in an archive that has gaps in history, the command will automatically fail, but can be run with warnings if a flag is provided.
- Graceful handling of HTTP errors with backoff retry logic (though graceful handling of large bucket file errors is coming soon)
- 0 length files identified as invalid

The tool is built around async streaming IO. This makes it quite fast, as we can perform streaming verification and writes of large files. The bulk of the logic is in `pipeline.rs`. This struct is responsible for managing concurrent workers, consuming history files, and in turn dispatching file fetching/processing work. All archival operations share much of the same work, which is handled by the pipeline:

1. Get upper and lower bound checkpoints to mirror/scan
2. Download history files and parse them to discover what files they reference
3. Dedup buckets files
4. Do work on files referenced by the history files

Each operation has different "do work" logic, which is handled by the `mirror_operation.rs` and `scan_operation.rs`. The pipeline hands the operation an open stream to a file, and they decide what to do with it and handle any errors. 

The following features will be coming soon and are outside the scope of this PR:
- repair
- support for `--verify` and `--thorough`
- JSON summaries of scan/mirror
- Error handling when streaming large buckets
- dump-xdr
- `--recent`
- s3 backends
- Quorum support, where instead of passing in a single source URL, a quorum config can be provided

This is my first Rust project, so I'd greatly appreciate comments on the tool itself, but also just Rust tips and feedback in general.

leighmcculloch commented:
I can't leave feedback on the PR because of the size of the PR it is crashing my browser. Can I suggest breaking the `testdata` addition out into a separate PR that can be merged without needing to use the GH interface, then we'll be able to provide some review comments directly on the CLI code. Otherwise I can leave feedback here in a top-level comment, but it will be harder to make sense of it without the comment appearing inline in the code.

SirTyson commented:
> I can't leave feedback on the PR because of the size of the PR it is crashing my browser. Can I suggest breaking the `testdata` addition out into a separate PR that can be merged without needing to use the GH interface to review it, then we'll be able to provide some review comments directly on the CLI code. Otherwise I can leave feedback here in a top-level comment, but it will be harder to make sense of it without the comment appearing inline in the code.

Sorry about that. It's still probably bigger than ideal, but I've removed the large test archive.

leighmcculloch commented on src/cli/mod.rs:
```diff
@@ -0,0 +1,125 @@
+
+#[derive(Subcommand, Debug)]
+pub enum Commands {
+    /// Mirror files from source archive to destination
+    Mirror(mirror::MirrorCmd),
```
Fyi, up to you, but the comment on the Mirror enum here, can also be placed on the MirrorCmd, which has the advantage that all the help docs for the MirrorCmd, both the description and the options, will appear in the same place. Mentioning because this wasn't obvious to me as possible and I only found this out late in the stellar-cli's dev.

leighmcculloch commented on src/cli/mod.rs:
```diff
@@ -0,0 +1,125 @@
+    Scan(scan::ScanCmd),
+}
+
+/// Global arguments shared by all commands
+#[derive(Debug, Clone)]
+pub struct GlobalArgs {
+    pub concurrency: usize,
+    pub skip_optional: bool,
+    pub max_retries: u32,
+    pub initial_backoff_ms: u64,
+}
```
Fyi instead of creating a GlobalArgs that you have to copy the options into, you can make this struct the definition of the options, and then specify the GlobalArgs as a single field on the Cmd, and add the `#[arg(flatten)]` so that the options flatten into the command.

For example: https://github.com/stellar/stellar-cli/blob/9d9f4b3464e2a75d34898e9a55fb28e3757d06cb/cmd/soroban-cli/src/commands/mod.rs#L78-L79

leighmcculloch commented on src/cli/scan.rs:
```diff
@@ -0,0 +1,64 @@
+        pipeline
+            .run()
+            .await
+            .map_err(utils::map_pipeline_error)
+            .map_err(Error::from)?;
```
Unless I'm missing something the `Error::from` shouldn't be required, because that's what the `?` will do for you automatically.
```suggestion
            .map_err(utils::map_pipeline_error)?;
```

leighmcculloch commented on src/utils.rs:
```diff
@@ -30,6 +31,16 @@ pub enum Error {
 }
 
+/// Helper function to map pipeline errors to library errors
+pub fn map_pipeline_error(err: pipeline::Error) -> crate::Error {
+    match err {
+        pipeline::Error::ScanOperation(scan_err) => crate::Error::ScanOperation(scan_err),
+        pipeline::Error::MirrorOperation(mirror_err) => crate::Error::MirrorOperation(mirror_err),
+        pipeline::Error::Io(io_err) => crate::Error::Io(io_err),
+        other => crate::Error::Other(other.to_string()),
+    }
+}
```
Fyi if you move this logic into a `impl From<pipeline::Error> for crate::Error` then all of the places like `.map_err(map_pipeline_error)` that it gets used should be replaceable with `?`.

leighmcculloch commented on src/utils.rs:
```diff
@@ -0,0 +1,281 @@
+    let current_ledger = state.current_ledger;
+    let highest_source_checkpoint = history_format::round_to_lower_checkpoint(current_ledger);
+
+    info!(
+        "Source archive reports current ledger: {} (checkpoint: 0x{:08x})",
+        current_ledger, highest_source_checkpoint
+    );
```
Not super important, but when playing with the command I appreciated in the trace output that the checkpoint is shown here in both integer and hex form. There are other logs where it's only shown in one or the other, sometimes in trace logs near each other, and it could be helpful to always log them both with one in brackets like this.

leighmcculloch commented on .github/workflows/rust.yml:
```diff
@@ -0,0 +1,78 @@
+    - run: cargo version
+    - run: cargo clippy --target ${{ matrix.sys.target }} --all-targets
+    - run: cargo build --target ${{ matrix.sys.target }}
+    # Tests can only run on native targets, not cross-compilation targets like aarch64-apple-darwin
+    - name: Run tests
+      if: matrix.sys.target != 'aarch64-apple-darwin'
+      run: cargo test --target ${{ matrix.sys.target }}
```
If you'd like to run the tests on all targets, change the x86 macos build to run on the `macos-13` runner. It's an Intel chip runner.

leighmcculloch reviewed (commented):
👏🏻 Looking great. Only one significant comment, the one about the `std::io::Error` usage. See inline.

jayz22 reviewed (changes requested):
Overall great work! I've tried the various commands they all seem to work nicely. Here is my feedback after reviewing the high level flow (I'm still processing some lower level details). 

The main thing I noticed is concurrency control parameters seem a bit inconsistent. One suggestion is to use a global tracking allocator for IO permit control, so that we would only need one concurrency parameter for number of checkpoint-processors.

SirTyson commented:
After much delay, I've revisited this PR. I've overhauled the pipeline after some suggestions from @jayz22 and @graydon, added more redundancy to network errors, and refactored + added many more unit tests. I'm experimented with AI on this update. Specifically, I spent a bunch of time writing detailed and fairly granular unit tests, with the hopes that this makes AI iteration more successful. 

A few notes to reviewers:

- This diff is quite large. Part of this is because it's the first PR in the repo, so it's setting up licenses, actions, etc. This also is a fairly robust MVP with concurrency and error redundancy. There are also a ton of tests. 
- I am not very experienced at rust and experimented quite a bit with AI tools, so paradigms might be weird an non-rusty.
- Unit tests don't pass, since it requires a test-data folder that broke github web UX since the diff was large. I'll push it in a follow up, but for now, trust me that unit tests pass locally.

leighmcculloch commented on README.md:
```diff
@@ -1,2 +1,95 @@
+## Description
+
+Stellar Archivist provides utilities to scan and mirror Stellar History Archives. It supports both HTTP/HTTPS and filesystem sources, allowing you to verify archive integrity and create local mirrors of remote archives.
+
```
````suggestion

## Install

From source:

```
cargo install --locked --features cli stellar-archivist
```
````

leighmcculloch commented on README.md:
```diff
@@ -1,2 +1,95 @@
+
+Stellar Archivist provides utilities to scan and mirror Stellar History Archives. It supports both HTTP/HTTPS and filesystem sources, allowing you to verify archive integrity and create local mirrors of remote archives.
+
+## Usage
```
```suggestion
## Usage (CLI)
```

leighmcculloch commented on README.md:
```diff
@@ -1,2 +1,95 @@
+- `--overwrite`: Overwrite existing files when mirroring
+- `--allow-mirror-gaps`: Allow creating gaps in destination archive
+
+## Architecture
```
```suggestion
## Usage (Lib)
```
